quantities/serialization/QuantityFactory.cs

```diff
@@ -0,0 +1,86 @@
+    private static readonly ConcurrentDictionary<Int32, IBuilder> complexCache = new();
+    private static readonly ConcurrentDictionary<QuantityModel, IBuilder> scalarCache = new();
+    private static readonly Type scalarVerification = typeof(TQuantity).MostDerivedOf<Dimensions.IDimension>();
+    private readonly IInject injector;
+    private readonly Type[] verifications;
+    public Int32 ExpectedModelCount => this.verifications.Length;
+    public Boolean IsScalarQuantity { get; }
+    private QuantityFactory(IInject injector, params Type[] verifications) : this()
+    {
+        this.injector = injector;
+        this.verifications = verifications;
+        this.IsScalarQuantity = injector is ScalarInjector;
+    }
+    public TQuantity Build(in Double value, in QuantityModel deserializedQuantity)
+    {
+        if (this.verifications.Length != 1) {
+            throw new InvalidOperationException($"Expected to build a quantity using '{this.verifications.Length}' sub models, but attempted to build with only one.");
+        }
+        var builder = Create(in deserializedQuantity, this.verifications[0], this.injector);
+        return TQuantity.Create(builder.Build(in value));
+    }
+    public TQuantity Build(in Double value, QuantityModel[] deserializedQuantity)
+    {
+        if (deserializedQuantity.Length != this.verifications.Length) {
+            throw new ArgumentOutOfRangeException(nameof(deserializedQuantity), $"Expected '{this.verifications.Length}' quantity sub models, but received only '{deserializedQuantity.Length}'.");
+        }
+        var builder = Create(deserializedQuantity, this.verifications, this.injector);
+        return TQuantity.Create(builder.Build(in value));
+    }
+
+    public static QuantityFactory<TQuantity> Create(String system) => system switch {
+        "frac" => new(Inject.Fractional, typeofQuantity.InnerTypes(typeof(IFraction<,>))),
+        "prod" => new(Inject.Multiplicative, typeofQuantity.InnerTypes(typeof(IProduct<,>))),
+        "square" => new(Inject.Square, typeofQuantity.InnerType(typeof(ISquare<>))),
+        "cubic" => new(Inject.Cubic, typeofQuantity.InnerType(typeof(ICubic<>))),
+        _ => new(Inject.Scalar, scalarVerification)
+    };
+
+    private static IBuilder Create(in QuantityModel model, Type verification, IInject injector)
+    {
+        if (scalarCache.TryGetValue(model, out var builder)) {
+            return builder;
+        }
+        return scalarCache[model] = ScalarBuilder.Create(in model, new TypeVerification(verification), injector);
+    }
+    private static IBuilder Create(QuantityModel[] models, Type[] verifications, IInject injector)
+    {
+        const Int32 offset = 2;
+        Int32 key = models.Length;
+        for (Int32 modelNumber = 0; modelNumber < models.Length; modelNumber++) {
+            unchecked {
+                key ^= (modelNumber + offset) * models[modelNumber].GetHashCode();
```

Comment:
I'm curious why you chose to use such a key, besides performance 🤩 Would the performance be greatly reduced if you would have used the fully localized quantity name (namespace+quantity name)?

Reply:
The issue here, is that we need to compute the key before we know the namespace of the quantity we're creating.
Like this we can check if we've deserialized this variant of "complex" quantity before we go through all the stuff involving reflection and just use a cached builder we know builds the type of quantity we want...

Reply:
What if you would serialize the unit in a fully determined way, and then directly create an instance of the quantity based on the fully determined namespace? Do you think this would be a way to avoid reflection altogether?

```json
{
  "length": {
    "value": 3.2,
    "si": {
      "prefix": "K",
      "unit": "m",
      "namespace": "Quantities.Length"
    }
  }
}
```

Reply:
I'd also like to get rid of reflection entirely. But alas, I don't see how that'd be effectively possible...

But let's take this up as a philosophical discussion at some point :-)